common/src/shared/express/types.ts

```diff
@@ -0,0 +1,42 @@
+import { Schema, Get } from '../api'
+import { PathParams } from 'express-serve-static-core'
+import express from 'express'
+
+type TypedHandlerForOperation<Operation extends Schema.Operation> = express.RequestHandler<
+  Get.PathParameters<Operation>,
+  Get.ResponseBody<Operation>,
+  Get.RequestBody<Operation>,
+  Get.QueryParameters<Operation>
+>
+
+export type TypedHandler<
+  API extends Schema.API,
+  PathName extends Get.PathName<API>,
+  OperationName extends Get.OperationName<API, PathName>
+> = TypedHandlerForOperation<Get.Operation<API, PathName, OperationName>>
+
+interface TypedMatcher<
+  API extends Schema.API,
+  ReturnType,
+  OperationName extends Schema.OperationName = Schema.OperationName
+> {
+  <PathName extends Get.PathName<API> & PathParams>(
+    path: PathName,
+    ...handlers: Array<
+      TypedHandler<API, PathName, OperationName & Get.OperationName<API, PathName>>
+    >
+  ): ReturnType
+}
+
+export interface TypedRouter<API extends Schema.API> {
+  all: TypedMatcher<API, this, 'all'>
+  get: TypedMatcher<API, this, 'get'>
+  post: TypedMatcher<API, this, 'post'>
+  put: TypedMatcher<API, this, 'put'>
+  delete: TypedMatcher<API, this, 'delete'>
+  patch: TypedMatcher<API, this, 'patch'>
+  options: TypedMatcher<API, this, 'options'>
+  head: TypedMatcher<API, this, 'head'>
+  // use: TypedMatcher<API, this>
+  // route: ??? (TODO)
```

Comment:
what's the TODO here?

Reply:
If you look at the `express.Router` type, most of the methods (`.all`, `.get`, `.post`, etc.) are covered here. This is deliberate; we're trying to have the `typexpress.Router` look, act, and quack like an Express router, so most of the Express router methods should be replicated here.

There is also `.use` which I have deliberately left out; it would be used for attaching child routers and other middleware.

`.route` is one of the remaining major methods in `Router`. I haven't figured out its type, but I also haven't figured out what it's used for. The TODO is to figure it out.